substrate/frame/meta-tx/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The overarching call type.
+		type RuntimeCall: Parameter
+			+ GetDispatchInfo
+			+ Dispatchable<
+				Info = DispatchInfo,
+				PostInfo = PostDispatchInfo,
+				RuntimeOrigin = Self::RuntimeOrigin,
+			> + IsType<<Self as frame_system::Config>::RuntimeCall>;
+		/// Signature type for meta transactions.
+		type Signature: Parameter + Verify<Signer = Self::PublicKey>;
+		/// Public key type used for signature verification.
+		type PublicKey: IdentifyAccount<AccountId = Self::AccountId>;
+		/// The context type of `Self::TxExtension`
+		type TxContext: Member + Default;
+		/// Transaction extension/s for meta transactions.
+		type TxExtension: TransactionExtension<<Self as Config>::RuntimeCall, Self::TxContext>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// TODO
+		TODO,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A call was dispatched.
+		Dispatched { result: DispatchResultWithPostInfo },
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Dispatch meta transaction.
+		///
+		/// origin must be signed
+		#[pallet::call_index(0)]
+		#[pallet::weight({
+			let dispatch_info = meta_tx.call.get_dispatch_info();
+			// TODO: plus T::WeightInfo::dispatch() which must include the weight of T::TxExtension
+			(
+				dispatch_info.weight,
+				dispatch_info.class,
+			)
+		})]
+		pub fn dispatch(origin: OriginFor<T>, meta_tx: MetaTxFor<T>) -> DispatchResultWithPostInfo {
+			let _who = ensure_signed(origin)?;
+
+			let (signer, signature) = match meta_tx.proof {
+				Proof::Signed(signer, signature) => (signer, signature),
+			};
+
+			let signed_payload = SignedPayloadFor::<T>::new(*meta_tx.call, meta_tx.tx_ext)
+				.map_err(|_| Error::<T>::TODO)?;
+
+			if !signed_payload.using_encoded(|payload| signature.verify(payload, &signer)) {
+				return Err(Error::<T>::TODO.into());
+			}
+
+			let origin = T::RuntimeOrigin::signed(signer);
+			let (call, tx_ext, _) = signed_payload.deconstruct();
+			let info = call.get_dispatch_info();
+			// TODO: to get the len we have to encode the original `meta_tx`.
+			let len = 0;
+			let mut ctx = T::TxContext::default();
+
+			let (_, val, origin) = T::TxExtension::validate(
+				&tx_ext,
+				origin,
+				&call,
+				&info,
+				len,
+				&mut ctx,
+				tx_ext.implicit().map_err(|_| Error::<T>::TODO)?,
+				&call,
+			)
+			.map_err(|_| Error::<T>::TODO)?;
+
+			let pre = T::TxExtension::prepare(tx_ext, val, &origin, &call, &info, len, &ctx)
+				.map_err(|_| Error::<T>::TODO)?;
+
+			let res = call.dispatch(origin);
+			let post_info = res.unwrap_or_else(|err| err.post_info);
+			let pd_res = res.map(|_| ()).map_err(|e| e.error);
+
+			T::TxExtension::post_dispatch(pre, &info, &post_info, len, &pd_res, &ctx)
```

Comment:
This would allow a graceful and harmless failure in cases like the one described [here](https://github.com/paritytech/polkadot-sdk/pull/4165#discussion_r1568226192) when a [`post_dispatch` fail is critical](https://github.com/paritytech/polkadot-sdk/blob/ba0f8de0c74b1308a53942c406137a748ef79925/substrate/primitives/runtime/src/traits.rs#L1576).